edx-solutions/xblock-drag-and-drop-v2
# OC-1320 DnDv2 - Allow to set automatic alignment of dragged items in drop zones

This PR adds an "align" field that it editable for each zone, with the options "none", "left", "right", and "center".  Align "none" corresponds to in the current default behaviour, where items are placed where the learner puts them via mouse, or centered in the zone via keyboard.

* Align "left" causes placed items to be stacked left to right in the zone, wrapping at the right edge of the zone.
* Align "right" works the same way as "left", but stacks from right to left.
* Align "center" causes placed items to be aligned in the center of the zone, but they are stacked top to bottom.

Items in "aligned" zones will not overlap each other, but if there's not enough space in the zone for all the items, they will overflow over the bottom of the zone, potentially overlapping other zones.

**JIRA tickets**: Implements OC-1320

**Discussions**:

1. The original specification requested that items placed in "aligned" zones be 
sequenced according the order they were placed by the learner.  This proved 
complex to implement, and so instead, items are sequenced in the order they are
configured by the instructor.

**Dependencies**: None

**Screenshots**:

Studio:
Updated DnDv2 Edit Zone screen:
![edit zone align](https://cloud.githubusercontent.com/assets/7556571/12798983/4cca04fc-cb1c-11e5-87df-e2f05cb7f995.png)

LMS:
Example with multiple items placed in zones with mixed alignment:
![mixed-alignment](https://cloud.githubusercontent.com/assets/7556571/12877175/fccdb7fa-ce5e-11e5-8876-307b50dde796.png)

Example with multiple items per zone, mixed alignment, with overlapping zones:
![overlapping zones](https://cloud.githubusercontent.com/assets/7556571/12877179/0718b2a0-ce5f-11e5-89f2-3b0ca31acd12.png)

Example with mixed alignment and numerical input:
![numerical input](https://cloud.githubusercontent.com/assets/7556571/12877185/2d36696e-ce5f-11e5-97c5-3fb3e40dbd52.png)

**Sandbox URLs**:

Studio: http://studio.master.dndv2.sandbox.opencraft.com/course/course-v1:OpenCraft+DnDv2+review
LMS: http://master.dndv2.sandbox.opencraft.com/courses/course-v1:OpenCraft+DnDv2+review/courseware/6fa41195e31b430398e755ca3bc61808/879aff4aa8c14ad385d896addcbd3d09/

**Partner information**: StGallen

**Merge deadline**: Thu Feb 4 EOD.  Apologies for the short notice!

**Testing instructions**:

1. Go to the course "Introduction to Drag and Drop v2" in Studio
2. In each Subsection, click Edit for the Drag and Drop exercise to see how the alignment is set up for each zone.

1. Go to the course "Introduction to Drag and Drop v2" in the LMS
2. In each Subsection, drag items to each zone to see how they are aligned.

*Expected behaviour*
* Alignment: none -- dropped items should remain where they're dropped (mouse), or be centered in the zone (keyboard)
* Alignment: left -- dropped items should be aligned with the left side of the zone.
* Alignment: center -- dropped items should be aligned in the center of the zone.
* Alignment: right -- dropped items should be aligned with the right side of the zone.
* Alignment: mixed -- items dropped in the top zone should be left-aligned, the middle zone should be center-aligned, and the bottom zone should be right-aligned.
* Multiple items per zone, mixed alignment -- Zone alignment should be as with the "Alignment: mixed" activity above.  
    * Items dropped in the top zone will sequence from left to right, wrapping at the right edge, and overflowing the bottom edge of the zone onto the zone below.
    * Items dropped in the middle zone will sequence from top to bottom, and will overflow the bottom edge of the zone onto the zone below.
    * Items dropped in the bottom zone will sequence from right to left, wrapping at the left edge, and overflowing the bottom edge of the zone onto the zone below.
* Multiple items per zone, mixed alignment, overlapping zones
    * The zones in this activity overlap, but item placement behaviour should be the same as with the "Multiple items per zone, mixed alignment" activity above. 
* Multiple items per zone, mixed alignment, numerical input -- Items in this activity have numerical input required.
   *  Item placement behaviour should be the same as with the "Multiple items per zone, mixed alignment" activity above.
   * Items dropped should not overlap Input fields and buttons.

**Author notes and concerns**:

1. The two tests that are failing are known to be flaky : `test_wide_image_mobile` and `test_square_image_mobile`

**Reviewers**
- [ ] Code: (@itsjeyd, @bradenmacdonald)
- [ ] Product: (TBD)
- [ ] Platform: (TBD)
- [ ] UX: (TBD)
- [ ] TNL: (TBD)
- [ ] a11y: (TBD)
- [ ] doc/strings: (TBD)

**Settings**

Basic XBlock settings:

```yaml
EDXAPP_FEATURES:
    ALLOW_ALL_ADVANCED_COMPONENTS: True
